src/tds/convert.c

```diff
@@ -1530,13 +1530,13 @@ tds_convert_real(const TDS_REAL* src, int desttype, CONV_RESULT * cr)
 		return sizeof(TDS_USMALLINT);
 		break;
 	case SYBINT4:
-		if (!IS_INT(the_value))
+		if (!IS_INT((double)the_value))
```

Comment:
nice catch, not sure however if this is really fixing an issue or just hiding a warning. It fixes loosing precision in the conversion but is it possible that it does not catch a number which could not be converted to int, or, otherwise, that it returns overflow for a number that can be converted to int?

Reply:
As https://godbolt.org/z/Ea18E5zTM illustrates, the issue is that due to limited `float` precision, you can't get too close to the limit without going slightly *over* it, at which point conversion will wrap around rather than saturating.

Reply:
It's more about rounding and precision. I was just wondering what's the best way without having to do conversions. Also I was looking at same function trying to understand if the same issue could affect other conversions.

Reply:
Good question.  Upon further consideration, how about leaving these call sites alone and instead tweaking the macros?  AFAICT,
```
#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) < (int64_t) TDS_INT_MAX + 1)
/* ... */
#define IS_UINT(x) (0 <= (x) && (x) < (uint64_t) TDS_UINT_MAX + 1)
```
should impose precisely correct limits.  They will let through `double` values exceeding `TDS_(U)INT_MAX` by less than one, but I'd argue that that's actually appropriate given that conversion is defined to round down anyway.

Reply:
Agreed. I would put all signed in the macros like
```c
#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) < (int64_t) TDS_INT_MAX + 1)
/* ... */
#define IS_UINT(x) (0 <= (x) && (x) < (int64_t) TDS_UINT_MAX + 1)
```
not sure about `int64_t`/`TDS_INT8`, maybe a future style update for the other parts

Reply:
I'm on board with the all-signed version.  64-bit types are trickier, particularly on the unsigned front, but AFAICT the only use of `IS_(U)INT8` is on a `long` (line 1116), where it amounts to a simple non-negativity check -- `tds_convert_real` and `tds_convert_flt8` use hand-coded boundary checks that attempt to test directly against `_MAX` values.  I'll prepare a revised patch.